Implemented Quad Tree

ZarakUofT opened the pull request:
This implementation helps reduce the search space when looking for nearest neighbors.

ZarakUofT committed:
Added Quadtrees

Implemented quad trees to decrease the search space when acquiring nearest nodes.

ZarakUofT committed:
Update index.html

MahmoodMahmood commented on quadtree.js:
```diff
@@ -0,0 +1,114 @@
+    //         strokeWeight(4)
+    //         node(p.x, p.y)
+    //     }
+    // }
```
Why comment this out? We can just leave it there and just not call it if we don't want it

MahmoodMahmood commented on quadtree.js:
```diff
@@ -0,0 +1,114 @@
+            return found; //empty array
+        }
+        if (this.divided) {
+            found = this.northeast.query(range, found);
+            found = this.southwest.query(range, found);
+            found = this.northwest.query(range, found);
+            found = this.southeast.query(range, found);
```
Pretty sure more than one of these could return, you need to `.concat` the result with the found array, not overwrite it.

MahmoodMahmood commented on quadtree.js:
```diff
@@ -0,0 +1,114 @@
+                if (range.contains(p)) {
+                    found.push(p);
+                }
+            }
```
consider using a combination of `.filter(` and `.concat(`

MahmoodMahmood commented on sketch.js:
```diff
@@ -52,31 +52,44 @@ class Tree {
+
   findNearestNode(x, y) {
     // https://stackoverflow.com/questions/8864430/compare-javascript-array-of-objects-to-get-min-max
-    return this.getAllNodes().reduce((prev, curr) => prev.dist(x,y) < curr.dist(x,y) ? prev : curr)
+    //return this.getAllNodes().reduce((prev, curr) => prev.dist(x, y) < curr.dist(x, y) ? prev : curr)
```
you should delete these lines

MahmoodMahmood commented on sketch.js:
```diff
@@ -52,31 +52,44 @@ class Tree {
     // https://stackoverflow.com/questions/8864430/compare-javascript-array-of-objects-to-get-min-max
-    return this.getAllNodes().reduce((prev, curr) => prev.dist(x,y) < curr.dist(x,y) ? prev : curr)
+    //return this.getAllNodes().reduce((prev, curr) => prev.dist(x, y) < curr.dist(x, y) ? prev : curr)
+    let radius = 50
+    for (let i = 1; i <= 5; i++) {
```
I'm not a fan of these hard coded values because we may want to change the size of the canvas (this also goes for my code, I was being lazy when I initially wrote it). I suggest passing in a max size to this function and say just doubling the current search radius until we find the nearest node or we hit the max (making sure we also check against the max radius itself).

MahmoodMahmood replied:
PS: I think you should start with a much smaller radius if you implement the doubling thing, something like 10 or so.

MahmoodMahmood commented on sketch.js:
```diff
@@ -127,21 +140,27 @@ class Obstacle {
+//Quad Tree initialization
+let cellPointLimit = 4
+let boundary = new Rectangle(canvas_width / 2, canvas_height / 2, canvas_width, canvas_height)
+let qtree = new Quad(boundary, cellPointLimit)
```
Should definitely be in the Tree class

MahmoodMahmood commented on sketch.js:
```diff
@@ -127,21 +140,27 @@ class Obstacle {
 let t = new Tree(100, 200)
 
+//Quad Tree initialization
+let cellPointLimit = 4
```
should be a `const`, can be hard coded in the Tree class

MahmoodMahmood reviewed (changes requested):
Well done overall, the code is very clean and most of the changes I am asking are minor.

ZarakUofT committed:
Changed the boundary around each node from rectangular to circular

While finding the nearest neighbours of a node, a rectangle was created around it and nodes were sought in it. Now, I have changed it to creating a circle around a node and searchign for the nearest node.

MahmoodMahmood reviewed (approved):
Awesome job, its so much faster now!
